truffe2/accounting_main/tests.py

```diff
@@ -1,3 +1,289 @@
+
+    def test_budget_mayi(self):
+        self.call_check_json('/accounting/main/budget/mayi')
+
+    def test_budget_json(self):
+        self.call_check_json('/accounting/main/budget/json')
+
+    def test_budget_deleted(self):
+        from accounting_main.models import Budget
+        Budget(id=2, name='bad budget', unit_id=1, accounting_year_id=1, costcenter_id=1, deleted=True).save()        
+        self.call_check_html('/accounting/main/budget/deleted', data={'upk':1})
+        self.call_check_redirect('/accounting/main/budget/deleted', method='post', data={'upk':1, 'pk':2}, redirect_url='/accounting/main/budget/')
+
+    def test_budget_logs(self):
+        self.call_check_html('/accounting/main/budget/logs')
+
+    def test_budget_logs_json(self):
+        self.call_check_json('/accounting/main/budget/logs/json')
+
+    def test_budget_add(self):
+        self.call_check_html('/accounting/main/budget/~/edit', data={'upk':1, 'ypk':1})
+        self.call_check_redirect('/accounting/main/budget/~/edit', method='post',
+                             data={'upk':1, 'ypk':1, 'tags':'', 'name':'my budget', 'costcenter':1}, redirect_url='/accounting/main/budget/2/')
+
+    def test_budget_edit(self):
+        self.call_check_html('/accounting/main/budget/1/edit')
+        self.call_check_redirect('/accounting/main/budget/1/edit', method='post',
+                                 data={'tags':'', 'name':'new budget', 'costcenter':1}, redirect_url='/accounting/main/budget/1/')
+
+    def test_budget_delete(self):
+        self.call_check_html('/accounting/main/budget/1/delete')
+
+    def test_budget_show(self):
+        self.call_check_html('/accounting/main/budget/1/')
+
+    def test_budget_switch_status(self):
+        self.call_check_text('/accounting/main/budget/1/switch_status', data={'dest_status':'2_treated', 'from_list':'from_list'})
+        self.call_check_text('/accounting/main/budget/1/switch_status?dest_status=2_treated&from_list=from_list', method='post', data={'do':'it'})
+        self.assertIn('window.location.reload();', self.response.content)
+
+    def test_budget_contact(self):
+        self.call_check_text('/accounting/main/budget/1/contact/agep_compta') 
+        self.call_check_text('/accounting/main/budget/1/contact/agep_compta', method='post', data={'key':'agep_compta', 'subject':'ask', 'message':'blabla', 'receive_copy':True}) 
+
+    def test_accountingline_list(self):
+        self.call_check_html('/accounting/main/accountingline/', data={'iSortingCols':2, 'iSortCol_0':1, 'iSortCol_1':3, 'iDisplayLength':5, 'iDisplayStart':0, 'sSearch': 'account'})
+
+    def test_accountingline_mayi(self):
+        self.call_check_json('/accounting/main/accountingline/mayi')
+
+    def test_accountingline_json(self):
+        self.call_check_json('/accounting/main/accountingline/json')
+
+    def test_accountingline_deleted(self):
+        from accounting_main.models import AccountingLine
```

Comment:
What should motivate someone to choose between TopLevenl and that kind of import ?  @the-glu

Reply:
Top-level imports are the norm, one should really avoid in-place imports since they nuance the readability of the code, concerns and expectations: As a developer, when I open a file, I can directly see what this file depends on and make myself a great idea of what this file does. Moreover, inline imports are not necessarily executed in all codepaths of the file, which may lead to unexpected errors at runtime that could have been avoided.